### PR TITLE
NAS-107954 / 20.10 / middlewared: properly close fds and report errors when changing system dataset

### PR DESCRIPTION
**Pull request by william-gr**

### CODE REVIEW

File: src/middlewared/middlewared/service.py [hunk 1]
```diff
@@ -1097,15 +1104,17 @@ def reconfigure_logging(self):
...
+        self.__kill_multiprocessing()
 
         self.middleware.send_event('core.reconfigure_logging', 'CHANGED')
 
+    @private
+    def stop_logging(self):
+        stop_logging()
+        self.__kill_multiprocessing()
+
+        self.middleware.send_event('core.reconfigure_logging', 'CHANGED', fields={'stop': True})
```
Comment (themylogin) [inline src/middlewared/middlewared/service.py:1116]: Why don't you just make a `core.stop_logging` event?

File: src/middlewared/middlewared/plugins/sysdataset.py [hunk 2]
```diff
@@ -416,6 +416,9 @@ async def migrate(self, _from, _to):
             path = '/tmp/system.new'
             if not os.path.exists('/tmp/system.new'):
                 os.mkdir('/tmp/system.new')
+            else:
+                # Make sure we clean up any previous attempts
+                await run('umount', '-R', path, check=False)
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/sysdataset.py:421]: Does FreeBSD have this option?

File: src/middlewared/middlewared/plugins/sysdataset.py [hunk 3]
```diff
@@ -344,8 +344,8 @@ async def __umount(self, pool, uuid):
         for dataset, name in reversed(self.__get_datasets(pool, uuid)):
             try:
                 await run('umount', '-f', dataset)
-            except subprocess.CalledProcessError:
-                self.logger.warning('Unable to umount %r', dataset)
+            except subprocess.CalledProcessError as e:
+                raise CallError(f'Unable to umount {dataset}: {e.stderr.decode()}')
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/sysdataset.py:348]: I remember we were not able to unmount something sometimes but it was fine to ignore it, did it happen to you while working on this issue?

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/service.py (hunk 1 above)
Comment (william-gr) [reply src/middlewared/middlewared/service.py:1116]: Thought about it but just make things harder for no reason.

File: src/middlewared/middlewared/plugins/sysdataset.py (hunk 2 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/sysdataset.py:421]: Dunno, also not relevant anymore

File: src/middlewared/middlewared/plugins/sysdataset.py (hunk 3 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/sysdataset.py:348]: It was not fine, we were just lazy to handle it correctly. In Linux -f does not really force anything.

File: src/middlewared/middlewared/plugins/sysdataset.py (hunk 2 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/sysdataset.py:421]: Does that mean we don't need FreeBSD code in `master` anymore?

**Pull request merged**